components/kserve/kserve.go

```diff
@@ -187,6 +182,12 @@ func (k *Kserve) configureServerless(instance *dsciv1.DSCInitializationSpec) err
 			return fmt.Errorf("ServiceMesh is need to set to 'Managed' in DSCI CR, it is required by KServe serving field")
 		}
 
+		// check on dependent operators if all installed in cluster
+		dependOpsErrors := checkDependentOperators(cli).ErrorOrNil()
+		if dependOpsErrors != nil {
+			return dependOpsErrors
+		}
```

Comment:
@israel-hdez Aren't we checking this in a precondition anyway? 

https://github.com/opendatahub-io/opendatahub-operator/blob/ae3b539fb64829a1e1a8854e9ccefd895a5fb752/components/kserve/serverless_setup.go#L11-L28

This check is called as part of `serverlessFeatures.Apply();` below, so I wonder if there's a bug in preconditions or this change is really fixing the problem.

Reply:
The pre-condition is checked only when the state is [Managed](https://github.com/opendatahub-io/opendatahub-operator/blob/incubation/components/kserve/kserve.go#L184) .

This PR fixes the issue when state is set to `Unmanaged` and kserve controller pods are not deployed due to operator [prereqs](https://github.com/opendatahub-io/opendatahub-operator/blob/incubation/components/kserve/kserve.go#L112)

Reply:
the `PreConditions` and the `OperatorExists` (called by `checkDependentOperators/checkRequiredOperatorsInstalled`) are similar but still a bit different
`PreConditions` checks the serving CRD if exist in this case
`OperatorExists` checks if both 2 operators serving and servicemesh are installed.

But now i start to think if this is correct.

say we have the `checkDependentOperators` fall in the condition of:
serving is Managed + ServiceMesh is Unamaged or Removed ? 

maybe the correct way is to make this checkDependentOperators() take operator name as parameter, so in this case `checkDependentOperators(ServiceMeshOperator)` only?

Reply:
> The pre-condition is checked only when the state is [Managed](https://github.com/opendatahub-io/opendatahub-operator/blob/incubation/components/kserve/kserve.go?rgh-link-date=2024-02-11T23%3A47%3A28Z#L184) .
> 
> This PR fixes the issue when state is set to `Unmanaged` and kserve controller pods are not deployed due to operator [prereqs](https://github.com/opendatahub-io/opendatahub-operator/blob/incubation/components/kserve/kserve.go?rgh-link-date=2024-02-11T23%3A47%3A28Z#L112)

I am a bit puzzled, maybe my coffee does not work yet, but this is the same block of code where both calls are made

https://github.com/VaishnaviHire/opendatahub-operator/blob/2356d5430b716cb5d20f82e4ca2f31c79bcec70f/components/kserve/kserve.go#L179-L195

Reply:
> The pre-condition is checked only when the state is [Managed](https://github.com/opendatahub-io/opendatahub-operator/blob/incubation/components/kserve/kserve.go?rgh-link-date=2024-02-11T23%3A47%3A28Z#L184) .
> 
> This PR fixes the issue when state is set to `Unmanaged` and kserve controller pods are not deployed due to operator [prereqs](https://github.com/opendatahub-io/opendatahub-operator/blob/incubation/components/kserve/kserve.go?rgh-link-date=2024-02-11T23%3A47%3A28Z#L112)

So this is about checking if Service Mesh operator is configured in the cluster? Maybe I'm slow today and my coffee does not work yet :)

In such a case how is it supposed to work? I am not sure I understand, as from the first check: 

https://github.com/VaishnaviHire/opendatahub-operator/blob/2356d5430b716cb5d20f82e4ca2f31c79bcec70f/components/kserve/kserve.go#L179-L195

this newly added code will only be invoked if both `ServiceMesh` and `Serverless` are in `Managed` state.

Reply:
>say we have the checkDependentOperators fall in the condition of:
serving is Managed + ServiceMesh is Unamaged or Removed ?
maybe the correct way is to make this checkDependentOperators() take operator name as parameter, so in this case checkDependentOperators(ServiceMeshOperator) only?

i was wrong for the above comments.

Reply:
@VaishnaviHire thanks to @zdtsw (we had a chat about it) I understand now what is it supposed to unblock :) Coffee still does not seem to work.

However, I still wonder if we need this check moved in here. Maybe we should clean up preconditions instead.

Reply:
I think we need more clarity about [RHOAIENG-2951](https://issues.redhat.com/browse/RHOAIENG-2951). At the moment, KServe _requires_ OSSM and Serverless to be present in the cluster. To me, it makes sense to check presence of a valid setup of these two components, even for `Unmanged`, so not sure if the Jira is fully valid.

Maybe, the check is what needs to be fixed. I don't think it should be conditioned.